File: sdks/python/trace_harness/view/display.py
"""DisplayNode —— 渲染产物：一棵脱离 ctx/kind 的显示树。

facet 不拼字符串、不碰格式：它只产出结构化的 DisplayNode（本行 = kind/name/brief，
加 engine 绑上来的 findings），由序列化器（text/markdown/html/treecli）各自成文。一个 facet
也可以产出**合成节点**（kind="" 的 collapse 行 / plan 分组头）——没有对应 Node 时 node_ids 为空。
"""

from __future__ import annotations

from dataclasses import dataclass, field
from math import ceil
from typing import Protocol, runtime_checkable

from trace_harness.model.node import Field, Finding


@runtime_checkable
class Compact(Protocol):
    """Choose a name for a target ratio and report its actual ratio.

    Serializers use ``actual`` as the next breakpoint, so implementations should return the
    highest-fidelity stable projection that fits ``expect``.
    """

    def compact(self, expect: float) -> tuple[str, float]: ...


def _name_length(value: str) -> int:
    return sum(2 if ord(character) > 255 else 1 for character in value)


@dataclass(frozen=True)
class DisplayName:
    name: str
    detail: str = ""

    def compact(self, expect: float) -> tuple[str, float]:
        candidates = (
            (f"{self.name} · {self.detail}", self.detail, self.name)
            if self.detail
            else (self.name,)
        )
        raw_length = max(1, _name_length(candidates[0]))
        projections = tuple(
            (name, _name_length(name) / raw_length) for name in dict.fromkeys(candidates)
        )
        target = max(0.0, min(1.0, expect))
        return next(
            (projection for projection in projections if projection[1] <= target),
            min(projections, key=lambda projection: projection[1]),
        )


def name_projections(named: DisplayNode | Compact, default_name: str = "") -> list[str]:
    """Serialize every distinct name needed by an integer display-width budget."""
    owner = named if isinstance(named, Compact) else DisplayName(default_name or named.name)
    raw, actual = owner.compact(1)
    raw_length = max(1, _name_length(raw))
    projections = [raw]
    budget = ceil(actual * raw_length) - 1
    while budget >= 0:
        expect = budget / raw_length
        name, actual = owner.compact(expect)
        if name not in projections:
            projections.append(name)
        if actual > expect:
            break
        # One projection covers every integer budget down to its actual ratio.
        budget = min(budget - 1, ceil(actual * raw_length) - 1)
    return projections


@dataclass
class DisplayNode:
    kind: str  # "" = 合成行（折叠摘要 / 分组头），序列化时不打 `name` 反引号
    name: str
    brief: list[Field] = field(default_factory=list)
    node_ids: list[str] = field(default_factory=list)  # 本行代表的 Node；合成行为空，聚合行为 N 个
    children: list[DisplayNode] = field(default_factory=list)
    findings: list[Finding] = field(default_factory=list)  # engine 按 node_id 绑定（含上浮的）
    folded: int = 0  # 本行折叠/聚合掉多少个原生 Node（× N / rollup 提示）

File: sdks/typescript/trace-harness/src/view/display.ts
import type { Field, Finding } from "../model/node";

export interface Compact {
  // expect is a target upper bound; implementations own their semantic views
  // and may return an actual ratio well below it. Serializers use actual as the
  // next breakpoint, so it must describe the returned projection truthfully.
  compact(expect: number): readonly [name: string, actual: number];
}

function nameLength(value: string): number {
  return [...value].reduce((length, character) => (
    length + (character.codePointAt(0)! > 255 ? 2 : 1)
  ), 0);
}

function clampRatio(value: number): number {
  return Math.max(0, Math.min(1, value));
}

export class DisplayName implements Compact {
  constructor(
    readonly name: string,
    readonly detail = "",
  ) {}

  compact(expect: number): readonly [name: string, actual: number] {
    const candidates = this.detail
      ? [`${this.name} · ${this.detail}`, this.detail, this.name]
      : [this.name];
    const rawLength = Math.max(1, nameLength(candidates[0]!));
    const projections = [...new Set(candidates)].map((name) => (
      [name, nameLength(name) / rawLength] as const
    ));
    return projections.find(([, actual]) => actual <= clampRatio(expect))
      ?? projections.reduce((shortest, projection) => (
        projection[1] < shortest[1] ? projection : shortest
      ));
  }
}

export interface DisplayNode extends Partial<Compact> {
  kind: string;
  name: string;
  brief: Field[];
  node_ids: string[];
  children: DisplayNode[];
  findings: Finding[];
  folded: number;
}

export function nameProjections(named: DisplayNode | Compact, defaultName = ""): string[] {
  const owner: Compact = typeof named.compact === "function"
    ? named as Compact
    : new DisplayName(defaultName || (named as DisplayNode).name);
  const compact = (expect: number) => owner.compact(expect);
  const [raw, rawActual] = compact(1);
  const rawLength = Math.max(1, nameLength(raw));
  const projections = [raw];
  // Offline HTML cannot call the original object. Use each returned ratio as the
  // next breakpoint instead of probing every display-column budget.
  let budget = Math.ceil(rawActual * rawLength) - 1;
  while (budget >= 0) {
    const expect = budget / rawLength;
    const [name, actual] = compact(expect);
    if (!projections.includes(name)) projections.push(name);
    if (actual > expect) break;
    budget = Math.min(budget - 1, Math.ceil(actual * rawLength) - 1);
  }
  return projections;
}
